Improve state models and measurement models

xEnVrE opened the pull request:
### Within this PR:
#### State models
- New hierarchy for state models:
   - `StateModel` represents a generic `x_k = f(x_{k-1}, u_{k-1}, w_{k-1})`
   - Implemented `AdditiveStateModel` representing `x_k = f(x_{k-1}, u_{k-1}) + w_{k-1}` with `w` white Gaussian noise
   - Implemented `LinearStateModel` inheriting from `AdditiveStateModel`, representing `x_k = F_{k-1} x_{k-1} + w_{k-1}`
   - Implemented `LTIStateModel` inheriting from `LinearStateModel`, representing `x_k = F x_{k-1} + w_{k-1}`
- Removed `StateModel` methods `getNoiseCovarianceMatrix` and `getNoiseSample`
- Added non-pure virtual `StateModel` methods `getJacobian` and `transitionProability`
- Class `WhiteNoiseAcceleration` now inherits from `LinearStateModel`
- Class `SimulatedStateModel` now supports state vectors of any size (before was 4)

#### Measurement models
- New hierarchy for measurement models:
  - `MeasurementModel` represents a generic `y_k = h(x_k, r_k)` with `r_k` Gaussian noise
  - Implemented `LinearMeasurementModel` inheriting from `MeasurementModel`, representing `y_k = H_k x_k + r_k`
  - Implemented `LTIMeasurementModel` inheriting from `LinearMeasurementModel`, representing `y_k = H x_k + r_k`
- Removed `MeasurementModel` method `getNoiseSample`
- Class `LinearModel` now inherits from `LinearMeasurementModel`
-  Method `MeasurementModel::measure` replaces method `MeasurementModel::getAgentMeasurements` and **does not take the state as input**
- Renamed method `MeasurementModel::bufferAgentData` to `MeasurementModel::freezeMeasurements` to better explain its meaning

After these changes `LinearSensor` looses its method `measure` that is implemented in inheriting classes. E.g., `SimulatedLinearSensor` implements `measure`.

Also, from now on it is intended that the method `measure` can be called any number of time within the same time step (there are several algorithms where this happen). For this reason the actual computation of the measurement should happen in `freezeMeasurements` while `measure` just returns the last frozen measurement. E.g., `SimulatedLinearSensor` now effectively simulate measurements in `freezeMeasurements`. 

#### Tests
- Updated test_SIS_Decorators (since use `MeasurementModelDecorator` inheriting from `MeasurementModel`)

#### Version
- Increase version to 0.9.101 since breaking API

#### Changelog
- Update CHANGELOG.md

xEnVrE committed:
Implement class LinearStateModel

xEnVrE committed:
Add method getJacobian to class StateModel

xEnVrE committed:
Remove method getNoiseSample from class MeasurementModel

xEnVrE committed:
Remove method getNoiseSample from class MeasurementModelDecorator

xEnVrE committed:
Method LinearModel::getNoiseSample do not override

xEnVrE committed:
Method MeasurementModel::measure replaces method MeasurementModel::getAgentMeasurements

xEnVrE committed:
Method MeasurementModelDecorator::measure substitues method getAgentMesurements

xEnVrE committed:
Use MeasurementModel::measure instead of MeasurementModel::getAgentMeasurements in GaussianLikelihood::likelihood

xEnVrE committed:
Class LinearModel do not implement MeasurementModel::measure

This method is implemented in classes inheriting from LinearModel

xEnVrE committed:
Method SimulatedLinearSensor::measure substitutes method getAgentMeasurements

xEnVrE committed:
Add class LinearMeasurementModel

xEnVrE committed:
Add class LTIMeasurementModel

xEnVrE committed:
Implement class AdditiveStateModel

xEnVrE committed:
Remove method StateModel::getNoiseSample from the class StateModel

xEnVrE committed:
Remove method getNoiseSample from the class StateModelDecorator

xEnVrE committed:
WhiteNoiseAcceleration::getNoiseSample do not override

xEnVrE committed:
Remove method getNoiseCovarianceMatrix from class StateModel

xEnVrE committed:
Remove method getNoiseCovarianceMatrix from class StateModelDecorator

xEnVrE committed:
WhiteNoiseAcceleration::getNoiseCovarianceMatrix do not override

xEnVrE committed:
Implement class LTIStateModel

xEnVrE committed:
WhiteNoiseAcceleration inherits from LinearStateModel

xEnVrE committed:
Update test_SIS_Decorators

xEnVrE committed:
Method MeasurementModel::freezeMeasurements substitutes method bufferAgentData const

This name better explains what the method does

The method is non-const since it will tipically store the measurement in a class member

xEnVrE committed:
Method MeasurementModelDecorator::freezeMeasurements subsistutes method bufferAgentData const

xEnVrE committed:
Method SimulatedLinearSensor::freezeMeasurements subsistutes method bufferAgentData const

xEnVrE committed:
Use method MeasurementModel::freezeMeasurements in UpdateParticles::correctStep

xEnVrE committed:
Move actual measurement computation from SimulatedLinearSensor::measure to freezeMeasurements

Rationale: if the method measure is called multiple times, within the same time step, the same measurement is returned

xEnVrE committed:
Class LinearModel inherits from LinearMeasurementModel

xEnVrE committed:
Class SimulatedStateModel support state vector of any size

xEnVrE committed:
Add non pure method StateModel::transitionProbability

xEnVrE committed:
Update version to 0.9.101

xEnVrE committed:
Update CHANGELOG.md

claudiofantacci commented on src/BayesFilters/src/AdditiveStateModel.cpp:
```diff
@@ -0,0 +1,17 @@
+{
+    /* FIXME
+       Add additive noise generation. */
+    propagate(cur_states, mot_states); // + add noise
```
Are we missing something here?

xEnVrE replied:
I added a comment on the commit related to this file because I wanted to discuss with you about it but probably isn't visible. A part from this, the code that is missing is already available in class `WhiteNoiseAcceleration`.

Since in this pull request `WhiteNoiseAcceleration` will inherits from `LinearStateModel` that in turn inherits from `AdditiveStateModel`, maybe we can move noise generation here once for all in order to have it for all the additive state models.

What do you think?

claudiofantacci replied:
> I added a comment on the commit related to this file because I wanted to discuss with you about it but probably isn't visible.

Unofrtunately, I don't see any comment 🤷‍♂ Maybe something wen wrong!





> maybe we can move noise generation here once for all in order to have it for all the additive state models.

Yes this is totally feasible. If we do like that, we must also add a virtual method to allow noise implementation. Am I right?

xEnVrE replied:
The simplest way could be like

```
propagate(cur_states, mot_states);
mot_state += getNoiseSample();
```
where `getNoiseSample` is a pure virtual method in class `AdditiveStateModel`.

However, If we agree that the noise is always white and Gaussian we can also move the noise implementation inside `AdditiveStateModel` and let the user change it by overriding the non-pure virtual method `AdditiveStateModel::getNoiseSample`.

Maybe the first option is more clean. What do you think?

claudiofantacci replied:
> However, If we agree that the noise is always white and Gaussian we can also move the noise implementation inside `AdditiveStateModel` and let the user change it by overriding the non-pure virtual method `AdditiveStateModel::getNoiseSample`.

I'm tempted to agree on this, but at the same time I forsee many constraints. For example, having an implementation in such "base" class, we would be forced to add some kind of parameters to tune the noise that would imply to have a non-trivial constructor for the class.
I would go for the first option, with a pure virtual function. What do you think about it?

xEnVrE replied:
I agree, it is better to have a simple implementation in "base" classes.

claudiofantacci commented on src/BayesFilters/include/BayesFilters/WhiteNoiseAcceleration.h:
```diff
@@ -34,7 +34,7 @@ class bfl::WhiteNoiseAcceleration : public StateModel
     void motion(const Eigen::Ref<const Eigen::MatrixXf>& cur_states, Eigen::Ref<Eigen::MatrixXf> prop_states) override;
 
-    Eigen::MatrixXf getNoiseSample(const int num) override;
+    Eigen::MatrixXf getNoiseSample(const int num);
```
If it does not override, can we move it to `protected`?

xEnVrE replied:
Of course!

claudiofantacci replied:
Great! 👍

xEnVrE replied:
Now it is again with `override` since implements pure virtual method `AdditiveStateModel::getNoiseSample`.

claudiofantacci commented on src/BayesFilters/include/BayesFilters/WhiteNoiseAcceleration.h:
```diff
@@ -36,7 +36,7 @@ class bfl::WhiteNoiseAcceleration : public StateModel
     Eigen::MatrixXf getNoiseSample(const int num);
 
-    Eigen::MatrixXf getNoiseCovarianceMatrix() override;
+    Eigen::MatrixXf getNoiseCovarianceMatrix();
```
If it does not override, can we move it to `protected`?

xEnVrE replied:
At the end of this PR  `WhiteNoiseAcceleration` inherits from `LinearStateModel` that in turn inherits from `AdditiveStateModel`. Then the method is changed again to `override` in a next commit.

claudiofantacci commented on src/BayesFilters/include/BayesFilters/LTIStateModel.h:
```diff
@@ -0,0 +1,44 @@
+class bfl::LTIStateModel : public bfl::LinearStateModel
+{
+public:
+    LTIStateModel(Eigen::MatrixXf& transition_matrix,
```
I generally prefer to have `const Eigen::Ref<const Eigen::MatrixXf>&` so that user can also pass blocks and other expressions. Is there a particular reason not ot do so?

xEnVrE replied:
No there are not. I will switch to `Eigen::Ref` then.

claudiofantacci replied:
Ok! 👍

claudiofantacci commented on src/BayesFilters/include/BayesFilters/LTIStateModel.h:
```diff
@@ -0,0 +1,44 @@
+    virtual ~LTIStateModel() noexcept { };
+
+    void propagate(const Eigen::Ref<const Eigen::MatrixXf>& cur_states,
+                           Eigen::Ref<Eigen::MatrixXf> prop_states) override;
```
Put this line next to the previous one 😄

xEnVrE replied:
Done!

claudiofantacci commented on src/BayesFilters/include/BayesFilters/LinearModel.h:
```diff
@@ -38,7 +38,7 @@ class bfl::LinearModel : public MeasurementModel
     std::pair<bool, bfl::Data> innovation(const bfl::Data& predicted_measurements, const bfl::Data& measurements) const override;
 
-    std::pair<bool, Eigen::MatrixXf> getNoiseSample(const int num) const override;
+    std::pair<bool, Eigen::MatrixXf> getNoiseSample(const int num) const;
```
If it does not override, can we move it to `protected`?

xEnVrE replied:
Done!

claudiofantacci commented on src/BayesFilters/include/BayesFilters/LTIMeasurementModel.h:
```diff
@@ -0,0 +1,32 @@
+class bfl::LTIMeasurementModel : public bfl::LinearMeasurementModel
+{
+public:
+    LTIMeasurementModel(const Eigen::MatrixXf& measurement_matrix, const Eigen::MatrixXf& noise_covariance_matrix);
```
I generally prefer to have `const Eigen::Ref<const Eigen::MatrixXf>&` so that user can also pass blocks and other expressions. Is there a particular reason not ot do so?

xEnVrE replied:
Done!

claudiofantacci commented on src/BayesFilters/src/StateModel.cpp:
```diff
@@ -8,3 +8,9 @@ Eigen::MatrixXf StateModel::getJacobian()
 }
+
+
+Eigen::VectorXf StateModel::transitionProbability(const Eigen::Ref<const Eigen::MatrixXf>& prev_states, Eigen::Ref<Eigen::MatrixXf> cur_states)
```
Since we have `getJacobian` maybe we could call it `getTransitionProbability`. What do you think about it?

xEnVrE replied:
Agree on it.

claudiofantacci replied:
Perfect! 👍

claudiofantacci commented on src/BayesFilters/include/BayesFilters/SimulatedStateModel.h:
```diff
@@ -17,7 +17,7 @@ class bfl::SimulatedStateModel : public Agent, public Logger
 {
 public:
     SimulatedStateModel(std::unique_ptr<StateModel> state_model,
```
Let's have all the signatures in the header on one line for now 👍
(Yes, probably this was already multi-line. My bad! 😬)

claudiofantacci reviewed (changes requested):
Good work! Just some minor changes.

Please also rebase on devel 👍

claudiofantacci commented on CMakeLists.txt:
```diff
@@ -6,7 +6,7 @@ endif()
 project(BayesFilters
         LANGUAGES CXX
-        VERSION 0.8.101)
+        VERSION 0.9.101)
```
There is no need to update the minor version to `9`. This is the `devel` branch of the next version that will have minor version `8`. All the breaking changes will go under that number and it is compliant with our custom devel SemVer.

xEnVrE replied:
Done and also changed patch from `101` to `100` as discussed.

claudiofantacci commented on CHANGELOG.md:
```diff
@@ -1,5 +1,46 @@
 # 📜 BayesFilters changelog
 
+## Version 0.9.101
```
All the entries of this chapter must be moved in chapter `Version 0.8.101`.
Moreover, all the entries of `Version 0.7.101` must also be moved to `Version 0.8.101` because we are making breaking changes.
Finally, since we are moving to version `0.8`, devel version should just be `0.8.100`, i.e. we are not developing toward release version `0.8.1`.

xEnVrE replied:
Ok, now is way more clear. Thank you for the explanation.

claudiofantacci replied:
Here for that 👍

claudiofantacci replied:
[This](https://semver.org/) is the official page of SemVer. Have a look at it 😉 
For this project, I opted for a slightly different version detailed [here](https://github.com/robotology/bayes-filters-lib#%EF%B8%8F-about-versioning).

xEnVrE replied:
I made the changes to CHANGELOG.md and tried to "merge" things using the right categories.

However it seems that some classes/methods that were referred in the 0.7.101 do not exist anymore or changed their name:
- class `Process`
- class `GenericData` (this should be what now is called `Data`)
- typedefs `Eigen*Data`
- method `MeasurementModel::registerProcessData`
- method `MeasurementModel::getProcessMeasurements`

Do you think I can remove these things from the CHANGELOG?

xEnVrE replied:
Also, if something was created and then removed, like the class `Particle` can I change things like

```
- Added Particle and ParticleSet classes.
(...)
- Removed Particle class
```
to 
```
- Added ParticleSet class.
```
?

claudiofantacci replied:
Yes absolutely! 👍 
Just make the CHANGELOG consistent with the status of the project up to the very last commit 😃

xEnVrE replied:
Perfect :+1:

xEnVrE committed:
Fix indentation in LTIStateModel.h

xEnVrE committed:
Use Eigen::Ref in LTIStateModel ctor

xEnVrE committed:
LinearModel::getNoiseSample is protected

xEnVrE committed:
Use Eigen::Ref in LTIMeasurementModel ctor

xEnVrE committed:
Rename method StateModel::transitionProbability to getTransitionProbability

xEnVrE committed:
Change signature of SimulatedStateModel ctor so that it is entirely on one line

xEnVrE committed:
Add pure virtual method getNoiseSample to class AdditiveStateModel

xEnVrE committed:
Implement noise generation in method  AdditiveStateModel::motion

xEnVrE committed:
Remove method WhiteNoiseAcceleration::motion

The implementation is the same as in the parent class AdditiveStateModel::motion

xEnVrE committed:
Method WhiteNoiseAcceleration::getNoiseSample overrides

Required since parent class AdditiveStateModel has pure virtual method getNoiseSample

xEnVrE committed:
Use std::size_t instead of int in signature of method WhiteNoiseAcceleration::getNoiseSample

xEnVrE committed:
Fix CHANGELOG.md

xEnVrE committed:
Fix version in CMakeLists.txt

claudiofantacci commented on src/BayesFilters/include/BayesFilters/LTIStateModel.h:
```diff
@@ -19,7 +19,7 @@ class bfl::LTIStateModel : public bfl::LinearStateModel
 
     void propagate(const Eigen::Ref<const Eigen::MatrixXf>& cur_states,
-                           Eigen::Ref<Eigen::MatrixXf> prop_states) override;
+                   Eigen::Ref<Eigen::MatrixXf> prop_states) override;
```
In headers, definitions are all on one line 😃

claudiofantacci commented on src/BayesFilters/include/BayesFilters/LTIStateModel.h:
```diff
@@ -13,8 +13,8 @@ namespace bfl {
-    LTIStateModel(Eigen::MatrixXf& transition_matrix,
-                  Eigen::MatrixXf& noise_covariance_matrix);
+    LTIStateModel(const Eigen::Ref<const Eigen::MatrixXf>& transition_matrix,
+                  const Eigen::Ref<const Eigen::MatrixXf>& noise_covariance_matrix);
```
In headers, definitions are all on one line 😃

claudiofantacci commented on src/BayesFilters/src/LTIStateModel.cpp:
```diff
@@ -6,8 +6,8 @@ using namespace bfl;
-LTIStateModel::LTIStateModel(Eigen::MatrixXf& transition_matrix,
-                             Eigen::MatrixXf& noise_covariance_matrix) :
+LTIStateModel::LTIStateModel(const Ref<const MatrixXf>& transition_matrix,
+                             const Ref<const MatrixXf>& noise_covariance_matrix) :
```
In implementation, lines are splitted any if they have more than 2 parameters 😃

claudiofantacci commented on src/BayesFilters/src/LTIMeasurementModel.cpp:
```diff
@@ -8,8 +8,8 @@ using namespace Eigen;
-    const Eigen::MatrixXf& measurement_matrix,
-    const Eigen::MatrixXf& noise_covariance_matrix
+    const Ref<const MatrixXf>& measurement_matrix,
+    const Ref<const MatrixXf>& noise_covariance_matrix
```
In implementation, lines are splitted any if they have more than 2 parameters 😃

claudiofantacci reviewed (approved):
Great!
I just commented style issues here and there. Fix t hem only if you have time, otherwise let's go on with the other PR. I'd like to setup a clang format to avoid this kind of problems in the PR.

After rebasing and tests are passed, I'll merge the PR.

xEnVrE committed:
Fix code style

In implementation, signatures are splitted if they have more than 2 parameters.

In headers, declarations are all on one line.

xEnVrE committed:
Clean CHANGELOG.md

xEnVrE commented:
I cleaned a bit the CHANGELOG considering the differences between the actual `devel` and `master`. However I think that still many classes that were introduced in `devel` are not indicated.

Can we open an issue for that in order to fix it once for all in a separate PR?

A part from this, I'm ready to rebase!

claudiofantacci commented:
Thanks! Awesome work 🎉
